Gracefully terminate service endpoints

aditighag opened the pull request:
Objective:
Upon removal of a service backend, drain all the existing connections to the backend, and ensure that the backend is
not selected for new connections.

Implementation:
Backend selection in the datapath is done using `lb service` map lookups, and backend information is retrieved via `lb backend` maps. Earlier, we removed the backend entry from the maps as soon as it was deleted. This can break existing active connections to the backend for per-packet based load-balancing, or unconnected UDP, wherein backend information retrieval lookups fail, and connections are redirected to other backends.

Kubernetes allows users to specify `terminationGracefulPeriod` in the pod spec in order to allow server pods to gracefully terminate active connections. In k8s v1.20, a new endpoint condition called as `Terminating` [1] was added. When we receive an update for a terminating backend, we skip adding it to the service map so that the backends isn't selected for new connections. But we keep the entry in the backend and affinity maps so that active connections continue are not disrupted. Once the endpoint deletion event is received, the backend state is fully removed.

See commit description for more details.

[1] https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#terminating

Fixes : #14844 

**Release note**
```release-note
Support graceful termination for service load-balancing such that active connections don't break when endpoints are deleted. 
```

aditighag committed:
vagrant: Enable feature gate

The `EndpointConditions.Terminating` field is
enabled with the `EndpointSliceTerminatingCondition`
feature gate [1]. The field is used to support graceful
termination for service load-balancing.

[1] https://github.com/kubernetes/api/blob/master/discovery/v1/types.go#L133

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
test: Enable feature gate

The `EndpointConditions.Terminating` field is
enabled with the `EndpointSliceTerminatingCondition`
feature gate [1]. The field is used to support graceful
termination for service load-balancing.

[1] https://github.com/kubernetes/api/blob/master/discovery/v1/types.go#L133

Signed-off-by: Aditi Ghag <aditi@cilium.io>

borkmann commented on pkg/service/service.go:
```diff
@@ -787,7 +787,10 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
+		// The terminating backend won't be selected to serve new requests.
+		if !b.Terminating {
+			backends[b.String()] = b.ID
+		}
```
Does the above also work for Maglev LB map?

aditighag replied:
Discussed offline - deferred to follow-ups.

aditighag committed:
test: Validate graceful termination of service endpoints

The test deploys test client and server apps, whereby
the server pod closes the client connection gracefully
while terminating. The client pod then exits successfully.
Also, the test validates that the terminating endpoint doesn't
serve new connections.

The test is skipped on GKE as it requires enabling
an alpha feature `EndpointSliceTerminatingCondition`.
Although the feature gate wasn't available on an alpha
cluster either [1]. Alpha features aren't supported on
EKS so skipped running the test on that platform.

[1] https://cloud.google.com/kubernetes-engine/docs/how-to/creating-an-alpha-cluster

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
k8s/endpoints: Fix outdated code references

Reference source code file should suffice
to look up the EndpointConditions struct
definition. No need to specify line numbers
that can change with upgrades.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
k8s/endpoints: Read `Terminating` state for endpoints

The `EndpointConditions.Terminating` field indicates
whether an endpoint is terminating. Propagate the
terminating state so that the endpoint can be
gracefully removed such that existing connections
are drained.
When an endpoint is deleted, k8s api server sends
an update endpoint slice event for that event with
the `terminating` field set. Once the graceful
termination period is over, it then sends a delete
event.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
loadbalancer/watcher: Propagate terminating state for service backends

The terminating state is used to gracefully terminate service
backends so that existing connections are drained. Propagate
the terminating state for internal backend type that's used
to program service BPF maps.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service: Gracefully terminate service backends

Objective:
Upon removal of a service backend, drain all the existing
connections to the backend, and ensure that the backend is
not selected for new connections.

Implementation:
Backend selection in the datapath is done using `lb service`
map lookups, and backend information is retrieved via `lb backend`
maps. Earlier, we removed the backend entry from the maps as soon
as it was deleted. This can break existing active connections to
the backend for per-packet based load-balancing, or unconnected UDP,
where backend information retrieval lookups fail, and connections are
redirected to other backends.
Kubernetes allows users to specify a `terminationGracefulPeriod` in the
pod spec in order to allow server pods to gracefully terminate active
connections. In k8s v1.20, a new endpoint condition called as
`Terminating` [1] was added. When an endpoint is deleted, k8s api
server first sends an update endpoint slice event with the terminating
state set to true for the endpoint. This state is propagated to the
internal backend state that's used to program datapath BPF maps.
When we receive such a backend, we skip adding it to the service map so
that the backend isn't selected for new connections. But we keep the entry
in the backend and affinity maps so that active connections continue
are not disrupted. Once the graceful termination period is over, k8s
watcher received the event to delete the endpoint. At that point, the
backend is removed from the remaining maps.

[1] https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#terminating

Related: #14844
Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
service_test: Add test to validate terminating endpoints processing

- Tests whether upsert service handles terminating backends, whereby
terminating backends are not added to the service map, but are kept
in the backends and affinity maps.
- Tests restore operations where terminating backends are not removed
from the backends and affinity maps.
- Add assertions for Maglev map operations.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
docs: Document requirements for graceful termination feature

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
config: Add a feature flag for graceful termination feature

The k8s `EndpointSliceTerminatingCondition` is currently an alpha
feature. Hence, introduce a feature flag until the k8s
feature is promoted to stable.
The flag is enabled by default as the feature provides
a core service load-balancing functionality.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

borkmann commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -771,6 +771,10 @@ data:
 {{- end }}
 
+{{- if hasKey .Values "enable-k8s-terminating-endpoint" }}
+  enable-k8s-terminating-endpoint: {{ .Values.enableTerminatingEndpoint | quote }}
```
I think this one needs a  bit of rework, e.g. above should check `hasKey .Values "enableTerminatingEndpoint"` at min. We should probably also add the entry description to values.yaml. Cc @joestringer

aditighag replied:
Yes, you are right. I forgot to add that file with my changes to the git commit.

borkmann commented on pkg/service/service.go:
```diff
@@ -787,7 +787,11 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
 	for _, b := range svc.backends {
-		backends[b.String()] = b.ID
+		// Skip adding the terminating backend to the service map as it
+		// won't be selected to serve new requests.
```
Could you also extend the comment a bit in that the backend will still be present in the backends map etc, and the refcounting will take care of the final cleanup once graceful period is over? Bit more verbosity of the mechanics would help for future context.

borkmann commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1158,6 +1158,14 @@ The current Cilium kube-proxy replacement mode can also be introspected through
     KubeProxyReplacement:   Strict	[eth0 (DR)]
 
+Graceful Termination
+*********************
```
tiny nit: I think one '*' too much, not sure if it still keeps rendering docs then.

borkmann commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1158,6 +1158,14 @@ The current Cilium kube-proxy replacement mode can also be introspected through
+Cilium's eBPF kube-proxy replacement supports graceful termination of service
+endpoints deletion. The feature requires at least Kubernetes version 1.20, and
+the feature gate ``EndpointSliceTerminatingCondition`` needs to be enabled. The
+feature can be disabled with the configuration option ``enable-k8s-terminating-endpoint``.
```
Could you add more verbosity in here? I think that would help users with the setup, e.g. what's the default time for the grace period? Also description on what happens in datapath during that period wrt terminating backend. Maybe also link to K8s docs, etc would help.

borkmann commented on test/k8sT/Services.go:
```diff
@@ -1548,6 +1549,133 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+					return !res.WasSuccessful()
+				}, 2*time.Minute, time.Second).Should(BeTrue(), "%q can connect when it should not work \n", pod)
+			}
+			wg.Wait()
```
Could we also assert whether the backend gets eventually removed as expected after graceful period? Meaning, to ensure that agent internal refcounting etc is all still working as expected?

aditighag replied:
That's validated in unit tests as that's the rightful place for map clean-up assertions.

errordeveloper commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -771,6 +771,10 @@ data:
   vlan-bpf-bypass: {{ .Values.bpf.vlanBypass | join " " | quote }}
 {{- end }}
 
+{{- if hasKey .Values "enable-k8s-terminating-endpoint" }}
```
```suggestion
{{- if hasKey .Values "enableTerminatingEndpoint" }}
```

borkmann commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1158,6 +1158,20 @@ The current Cilium kube-proxy replacement mode can also be introspected through
+Cilium's eBPF kube-proxy replacement supports graceful termination of service
+endpoints deletion. The feature requires at least Kubernetes version 1.20, and
+the feature gate ``EndpointSliceTerminatingCondition`` needs to be enabled. The
+feature can be disabled with the configuration option ``enable-k8s-terminating-endpoint``.
```
nit, I would make a slight amendment here, for example:

[...] needs to be enabled. By default, the Cilium agent then detects such terminating Pod state. If desired, the feature can also be disabled with the configuration option ``--enable-k8s-terminating-endpoint=false``.

brb commented on pkg/k8s/endpoints.go:
```diff
@@ -164,7 +165,16 @@ func ParseEndpointSliceV1Beta1(ep *slim_discovery_v1beta1.EndpointSlice) (Endpoi
+				// Propagate the terminating endpoint state so that we can gracefully
+				// remove those endpoints.
+				// More details : vendor/k8s.io/api/discovery/v1beta1/types.go
+				if sub.Conditions.Terminating != nil && !*sub.Conditions.Terminating {
```
Aren't we adding non-ready endpoint if it is in non-terminating state?

aditighag replied:
No, we are not. There is a `continue` on the next line.

brb replied:
- We won't invoke `continue` if `--enable-k8s-terminating-endpoints=false` for non-ready endpoints. Is it OK?
- We won't invoke `continue` for non-ready terminating endpoints. Same question.

aditighag replied:
>We won't invoke continue if --enable-k8s-terminating-endpoints=false for non-ready endpoints. Is it OK?

Ah, yes! Need to fix the case where user explicitly disables the feature flag. 

>We won't invoke continue for non-ready terminating endpoints. Same question.

Yes, as the comment indicates, we want to propagate the terminating state.

Edit : Refactored the code a bit, and also added unit tests.

brb commented on test/k8sT/manifests/graceful-termination.yaml:
```diff
@@ -0,0 +1,47 @@
+    app: graceful-term-server
+spec:
+  containers:
+    - name: server
```
Do we actually need new pods to test the graceful termination? Maybe the existing testDS pods are enough? If not, then could you move the images into the cilium repo/org?

aditighag replied:
Yes, we do - 1) Test phases are organized based on logged events by server/client 2) The server pod performs actions based on `SIGTERM` signal.
Yes, I've moved the image. Was waiting until Jenkins test infra issue was resolved.

brb commented on test/k8sT/Services.go:
```diff
@@ -1548,6 +1549,132 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+		)
+
+		BeforeAll(func() {
+			if helpers.SkipK8sVersions("<1.20") {
```
Could we move this to the SkipContextIf condition?

aditighag replied:
I decided to keep it looking at other references in the code - it's useful to have the context why the test is being skipped.

brb commented on test/k8sT/manifests/demo-graceful-term.yaml:
```diff
@@ -0,0 +1,21 @@
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: testclient
```
Don't we have multiple similar pods already available?

aditighag replied:
Nope, they are replicated in every e2e test that need the demo pods. Anyway I'm going to fold this into the graceful-test yaml file.

brb commented on test/k8sT/Services.go:
```diff
@@ -1548,6 +1549,132 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+				"kubeProxyReplacement": "disabled",
+			})
+
+			testPodYAML = helpers.ManifestGet(kubectl.BasePath(), "demo-graceful-term.yaml")
```
Shouldn't this be in a separate It() block?

aditighag replied:
I presume you meant that the actual test logic `Checking if terminating service endpoint doesn't serve new connections` should be moved to a separate It() block? Deployment of test pods will still happen in the `BeforeAll` block, no? Yes, we can do that.

brb commented on pkg/service/service_test.go:
```diff
@@ -710,3 +714,93 @@ func (m *ManagerTestSuite) TestLocalRedirectServiceOverride(c *C) {
+	c.Assert(len(m.lbmap.AffinityMatch[uint16(id1)]), Equals, 0)
+}
+
+// Tests terminating backend entries are removed after service restore.
```
We need to test whether terminating backend IDs have been restored after the restart. Otherwise, we are risking to allocate already used IDs.

aditighag replied:
Discussed offline, I'll revisit the code path along with the unit test case.

brb commented on pkg/service/service_test.go:
```diff
@@ -710,3 +714,93 @@ func (m *ManagerTestSuite) TestLocalRedirectServiceOverride(c *C) {
+	err = m.svc.RestoreServices()
+	c.Assert(err, IsNil)
+
+	// (Non-terminating) Backends have been restored
```
Why don't we restore the terminating backends?

brb commented on test/helpers/kubectl.go:
```diff
@@ -1195,6 +1195,21 @@ func (kub *Kubectl) Logs(namespace string, pod string) *CmdRes {
 
+// LogsWithLabel returns a CmdRes with command output from the
+// execution of `kubectl logs --previous -l <label string> -n <namespace>`.
+func (kub *Kubectl) LogsWithLabel(namespace string, labelStr string) *CmdRes {
```
Nit: maybe let's add "previous" to the method name to avoid any possible confusion.

brb commented on test/k8sT/Services.go:
```diff
@@ -1548,6 +1549,155 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+		}
+
+		BeforeAll(func() {
+			if helpers.SkipK8sVersions("<1.20.0") {
```
(GH doesn't allow me to comment in my previous comment) I'd still move it to the ctx skip condition. You can add the comment there saying why we skip the test. Otherwise, it makes it more difficult already overbloated and poorly structured suite file (I'm talking about the whole `K8sService` suite).

brb commented on test/k8sT/manifests/graceful-termination.yaml:
```diff
@@ -0,0 +1,69 @@
+      imagePullPolicy: IfNotPresent
+      command: [ "/client", "graceful-term-svc.default.svc.cluster.local.:8081" ]
+---
+apiVersion: apps/v1
```
This is a duplicate of the `test/k8sT/manifests/demo_ds.yaml`.

aditighag replied:
No, it's not. It contains a bunch of other test apps. :( See my previous comment - https://github.com/cilium/cilium/pull/17716#discussion_r745769554.

brb replied:
I meant this particular object is defined in both files.

brb commented on test/k8sT/Services.go:
```diff
@@ -1548,6 +1549,155 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			wg               sync.WaitGroup
+		)
+
+		waitForServer := func() {
```
Could you move the helper functions to `service_helpers.go`, so that the file size can be reduced for the sake of readability? Also, you could make it more generic so that it can be reused.

brb commented on test/k8sT/manifests/graceful-termination.yaml:
```diff
@@ -0,0 +1,69 @@
+  selector:
+    app: graceful-term-server
+---
+apiVersion: v1
```
It would make reviewing easier if you could write a few comments saying what the server and the client does, and what output/behavior is used for testing. Otherwise, a reviewer needs to dive into the implementations of both.

aditighag replied:
There is a detailed description of the test apps in the commit description, and also in the e2e test inline. I don't see any description in other test yamls. I'll see what I can do more.

brb commented on test/k8sT/Services.go:
```diff
@@ -1548,6 +1549,155 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			}
+		}
+
+		terminateServiceEndpoint := func() {
```
Ditto.

brb commented on test/k8sT/Services.go:
```diff
@@ -1548,6 +1549,155 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			if helpers.SkipK8sVersions("<1.20.0") {
+				Skip(fmt.Sprintln("Graceful termination feature requires at least K8s version 1.20"))
+			}
+			DeployCiliumOptionsAndDNS(kubectl, ciliumFilename, map[string]string{
```
To reduce the suite runtime, have you considered moving the test case in the beginning where we test per-packet LB? Then the redeploying with KPR off is not needed.

aditighag replied:
Nice suggestion! Will look into that. Thanks!

aditighag replied:
> have you considered moving the test case in the beginning where we test per-packet LB? Then the redeploying with KPR off is not needed.

Well, we don't enable kpr later on. We set kpr to `strict` on 4.19 and later. Per-packet LB tests are skipped on these kernels.

aditighag committed:
testutils/mockmaps: Populate Maglev table in UpsertService

Emulate the real behavior - https://github.com/cilium/cilium/blob/v1.10/pkg/maps/lbmap/lbmap.go#L104.

Preparatory commit to validate terminating backends behavior
in the service_test.go.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
service/lbmap: Consider active backends while cleaning up service map entries

Service entry currently keeps a count of previous count of backend
entries so that when service backends are updated, the BPF service
map entries for old backends can be deleted.

However, we skip adding terminating backend entries to the services
BPF map so that it's not selected for new connections. As a result,
we need to keep track of only the active backends (aka non-terminating)
while cleaning up obsolete backend entries.

See PR #17716 for more details about how terminating backends are
processed for graceful removal.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service: Account for terminating backends during restore

Terminating backends are not added to the BPF services map, but
are kept in the affinity and backends map for gracefully
terminating active connections to these backends. See PR #17716
for more details.
As a result, when we restore service state after agent restart, we
need to account for the terminating backends so that they are not
prematurely removed. Specifically, we need to defer clean-up of
orphan backends and affinity matches until after the agent's sync
with kubernetes api server has finished. This allows agent to have
the up-to-date state about a service's terminating backends, if any.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
service_test: Tests updated restore logic

Deletion of orphan resources is moved from
the restore path, and deferred until kubernetes
sync is finished to account for terminating
backends. Update the corresponding tests accordingly.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

brb reviewed (approved):
:rocket:

aojea commented on pkg/k8s/endpoints.go:
```diff
@@ -213,17 +235,33 @@ func parseEndpointPortV1Beta1(port slim_discovery_v1beta1.EndpointPort) (string,
+			if option.Config.EnableK8sTerminatingEndpoint {
+				// Terminating indicates that the endpoint is getting terminated. A
+				// nil values indicates an unknown state. Ready is never true when
+				// an endpoint is terminating. Propagate the terminating endpoint
```
Kubernetes default behaviors allows graceful terminating, it never breaks established connections, doesn't matter the endpoints state.

This feature is meant to be used to allow sending NEW traffic to Terminating endpoints, "if there is no ready endpoint, then use the ones that are terminating AND SERVING"
```
	// serving is identical to ready except that it is set regardless of the
	// terminating state of endpoints. This condition should be set to true for
	// a ready endpoint that is terminating. If nil, consumers should defer to
	// the ready condition.
	// +optional
	Serving *bool
```

you should skip the endpoint is is terminating and not service

/cc @thockin
